function.dd

```diff
@@ -121,14 +121,61 @@ int foo();
 
 $(H4 $(LNAME2 pure-functions, Pure Functions))
 
-        $(P Pure functions are functions which cannot access global or static, mutable
-        state save through their arguments. This can enable optimizations based on the fact
-        that a pure function is guaranteed to mutate nothing which isn't passed to it,
+    $(P Pure functions are functions which cannot directly access global or static
+        mutable state. It guarantees that a pure function call won't modify any
+        invisible places from the caller side.
+    )
+
+    $(P Different from some other functional programming languages, D's pure
+        function allows modification of caller side state through its mutable
+        parameters.
+
+        ---
+        pure int foo(int[] arr) { arr[] += 1; return arr.length; }
+        int[] a = [1, 2, 3];
+        foo(a);
+        assert(a == [2, 3, 4]);
+        ---
+
+        The minimum guarantee of "pure" is called "weak purity". It's designed to
+        increase modularity of functions.
+    )
+
+    $(P To control mutations, D has the $(D immutable) type qualifier. If all of
+        a $(D pure) function's parameters are $(D immutable) or copied values
+        without any indirections, it can guarantee that the pure function has
```

Comment:
$(P To prevent mutation, D offers the $(D immutable) type qualifier. If all of a $(D pure) function's parameters are $(D immutable) or copied values without any indirections (e.g. $(D int)), the type system guarantees no side effects.)